File: src/parser.zig
const std = @import("std");
const types = @import("types.zig");
const rt = @import("rt.zig");
const get_intrinsic = rt.get_intrinsic;
const ArrayList = std.ArrayList;
const Token = types.Token;
const TokenType = types.TokenType;
const Node = types.Node;

const TokenizeError = error{ UnterminatedStringLiteral, InvalidCharacterLiteral, OutOfMemory };

fn tokenize(allocator: std.mem.Allocator, source: []const u8) TokenizeError!ArrayList(Token) {
    var tokens: ArrayList(Token) = .empty;
    var pos: usize = 0;

    while (pos < source.len) {
        const cur = source[pos];
        var len: usize = 1;

        if (std.ascii.isWhitespace(cur)) {
            pos += 1;
            continue;
        }

        const token: Token = switch (cur) {
            '(' => .{ .type = .LParen, .value = "(" },
            ')' => .{ .type = .RParen, .value = ")" },
            '+' => .{ .type = .Add, .value = "+" },
            '-' => .{ .type = .Sub, .value = "-" },
            '*' => .{ .type = .Mul, .value = "*" },
            '/' => .{ .type = .Div, .value = "/" },
            '[' => .{ .type = .LBracket, .value = "[" },
            ']' => .{ .type = .RBracket, .value = "]" },
            ',' => .{ .type = .Comma, .value = "," },
            '.' => b: {
                if (pos + 1 < source.len and source[pos + 1] == '.') {
                    len = 2;
                    break :b .{ .type = .DoubleDot, .value = ".." };
                }

                continue;
            },
            '\'' => b: {
                const start = pos;
                len = 1;

                if (start + 1 >= source.len) {
                    return TokenizeError.UnterminatedStringLiteral;
                }

                if (!std.ascii.isAlphanumeric(source[start + 1])) {
                    return TokenizeError.InvalidCharacterLiteral;
                }

                len += 1;

                if (start + len >= source.len or source[start + len] != '\'') {
                    return TokenizeError.InvalidCharacterLiteral;
                }

                const slice = source[start + 1 .. start + 2];
                len += 1;
                break :b .{ .type = .Char, .value = slice };
            },
            '0'...'9' => b: {
                const start = pos;
                len = 1;
                while (start + len < source.len and std.ascii.isDigit(source[start + len])) {
                    len += 1;
                }

                break :b .{ .type = .Number, .value = source[start .. start + len] };
            },
            'a'...'z', 'A'...'Z' => b: {
                const start = pos;
                len = 1;
                while (start + len < source.len and std.ascii.isAlphabetic(source[start + len])) {
                    len += 1;
                }

                break :b .{ .type = .Symbol, .value = source[start .. start + len] };
            },
            '"' => b: {
                const start = pos + 1;
                len = 1;
                while (pos + len < source.len and source[pos + len] != '"') {
                    len += 1;
                }

                const val = source[start .. pos + len];
                if (pos + len >= source.len) {
                    return TokenizeError.UnterminatedStringLiteral;
                }

                len += 1;

                break :b .{ .type = .String, .value = val };
            },
            else => .{ .type = .EOF, .value = "" },
        };

        pos += len;
        try tokens.append(allocator, token);
    }

    try tokens.append(allocator, .{ .type = .EOF, .value = "" });

    return tokens;
}

const ParserError = error{ UnexpectedToken, InvalidCharacter, Overflow, OutOfMemory };

pub const Parser = struct {
    allocator: std.mem.Allocator,
    ast: ArrayList(*Node),
    tokens: ArrayList(Token),
    pos: usize,

    pub fn init(allocator: std.mem.Allocator, source: []const u8) !Parser {
        const tokens = try tokenize(allocator, source);
        return .{
            .allocator = allocator,
            .ast = .empty,
            .tokens = tokens,
            .pos = 0,
        };
    }

    pub fn deinit(self: *Parser) void {
        self.tokens.deinit(self.allocator);
        self.ast.deinit(self.allocator);
    }

    pub fn parse(self: *Parser) !?*Node {
        if (self.peek().type == .EOF) return null;

        const node = try self.parse_expr(0);
        try self.ast.append(self.allocator, node);
        return node;
    }

    pub fn parse_expr(self: *Parser, prec: u8) !*Node {
        var left = try self.parse_null_deno();

        while (true) {
            const next = self.peek();
            // treat the next token as an infix operator if
            // the token looks like the start of an expression
            const next_pr = self.get_precedence(next);
            if (prec >= next_pr) break;

            left = try self.parse_left_deno(left);
        }

        return left;
    }

    fn parse_null_deno(self: *Parser) ParserError!*Node {
        // null denominator
        const current = self.consume();
        return switch (current.type) {
            .Number => try self.alloc_node(.{ .literal = try std.fmt.parseInt(i64, current.value, 10) }),
            .String => try self.alloc_node(.{ .string = current.value }),
            .Char => try self.alloc_node(.{ .char = current.value[0] }),
            .Symbol => {
                if (get_intrinsic(current.value)) |_| {
                    const args = try self.allocator.alloc(*Node, 0);
                    return try self.alloc_node(.{ .intrinsic = .{ .name = current.value, .args = args } });
                }

                return error.UnexpectedToken;
            },
            .LBracket => {
                // list or range list
                if (self.peek().type == .Number) {
                    const start = self.consume();

                    if (self.peek().type == .DoubleDot) {
                        _ = self.consume();
                        // past [<num>..
                        const end = try self.expect(.Number);
                        const start_v = try std.fmt.parseInt(i64, start.value, 10);
                        const end_v = try std.fmt.parseInt(i64, end.value, 10);

                        // 1 is the default step
                        var step: i64 = 1;
                        // if there is a comma after the range, we expect for the user
                        // to define a step
                        if (self.peek().type == .Comma) {
                            _ = self.consume();
                            const step_t = try self.expect(.Number);
                            step = try std.fmt.parseInt(i64, step_t.value, 10);
                        }

                        _ = try self.expect(.RBracket);

                        if (step <= 0) return error.InvalidCharacter;
                        const count = @divTrunc((end_v - start_v), step) + 1;
                        if (count <= 0) return error.InvalidCharacter;

                        var items = try self.allocator.alloc(i64, @intCast(count));
                        var value = start_v;
                        for (0..@intCast(count)) |i| {
                            items[i] = value;
                            value += step;
                        }

                        return try self.alloc_node(.{ .list = items });
                    } else {
                        var items: ArrayList(i64) = .empty;
                        const first = try std.fmt.parseInt(i64, start.value, 10);
                        try items.append(self.allocator, first);

                        while (self.peek().type != .RBracket) {
                            if (self.peek().type == .Comma) _ = self.consume();
                            if (self.peek().type == .RBracket) break; // ^this caused a lot of headache...

                            const value = try std.fmt.parseInt(i64, (try self.expect(.Number)).value, 10);
                            try items.append(self.allocator, value);
                        }

                        _ = try self.expect(.RBracket);
                        return try self.alloc_node(.{ .list = try items.toOwnedSlice(self.allocator) });
                    }
                }

                _ = try self.expect(.RBracket);
                return try self.alloc_node(.{ .list = try self.allocator.alloc(i64, 0) });
            },
            .LParen => {
                const next = self.peek();

                const is_op = switch (next.type) {
                    .Add, .Sub, .Mul, .Div => true,
                    .Symbol => get_intrinsic(next.value) == null,
                    else => false,
                };

                if (is_op) {
                    const optk = self.consume();
                    const arg = try self.parse_expr(0);

                    _ = try self.expect(.RParen);

                    return try self.alloc_node(.{ .partial = .{ .op = optk.value, .arg = arg } });
                } else {
                    const node = try self.parse_expr(0);

                    _ = try self.expect(.RParen);

                    return node;
                }
            },

            else => error.UnexpectedToken,
        };
    }

    fn parse_left_deno(self: *Parser, left: *Node) ParserError!*Node {
        // the only infix op is an application, left associative.
        const right = try self.parse_expr(10);
        return try self.alloc_node(.{ .application = .{ .func = left, .arg = right } });
    }

    fn peek(self: *Parser) Token {
        if (self.pos >= self.tokens.items.len) return .{ .type = .EOF, .value = "" };
        return self.tokens.items[self.pos];
    }

    fn consume(self: *Parser) Token {
        const prev = self.peek();
        if (prev.type != .EOF) self.pos += 1;
        return prev;
    }

    fn expect(self: *Parser, typ: TokenType) error{UnexpectedToken}!Token {
        const prev = self.consume();
        if (prev.type != typ) {
            std.debug.print("expected {}, got {}\n", .{ typ, prev.type });
            return error.UnexpectedToken;
        }
        return prev;
    }

    fn alloc_node(self: *Parser, node: Node) !*Node {
        const n = try self.allocator.create(Node);
        n.* = node;
        return n;
    }

    fn get_precedence(self: *Parser, token: Token) u8 {
        _ = self;

        return switch (token.type) {
            .Number, .String, .LParen, .LBracket => 10, // previous expression is being applied to these tokens as they start a new expr
            else => 0,
        };
    }
};

File: src/parser.rs
use std::cmp::PartialEq;
use crate::error::{LamError, LamResult};

#[derive(Debug, PartialEq)]
pub enum Token {
    LParen,
    RParen,
    LBracket,
    RBracket,
    Comma,
    Semi,
    Range,
    Spread,
    Lambda,
    Number(f64),
    String(String),
    Char(char),
    Symbol(String), /* identifier but with a fancy name */
}

#[derive(Debug, Clone)]
pub struct MatchArm {
    pub pattern: Pattern,
    pub guard: Option<Box<Node>>,
    pub body: Box<Node>,
}

#[derive(Debug, Clone)]
pub enum Pattern {
    Literal(f64),
    StringLit(String),
    Wildcard,
    Var(String),
    List(Vec<Pattern>, Option<String>),
}

#[derive(Debug, Clone)]
pub enum Node {
    Literal(f64),
    Atom(String),
    Partial { op: String, arg: Box<Node> },
    Application { func: Box<Node>, arg: Box<Node> },
    List(Vec<Node>),
    Let { name: String, value: Box<Node>, body: Box<Node> },
    If { cond: Box<Node>, then: Box<Node>, els: Box<Node> },
    FnDef { name: String, params: Vec<String>, body: Box<Node> },
    LambdaDef { params: Vec<String>, body: Box<Node> },
    Match { expr: Box<Node>, arms: Vec<MatchArm> },
    UseModule { path: String },
}

pub struct Parser {
    tokens: Vec<Token>,
    pos: usize,
}

impl Parser {
    pub fn new(input: &'_ str) -> Parser {
        let tokens = tokenize(input);

        Parser {
            tokens,
            pos: 0
        }
    }

    pub fn parse(&mut self) -> LamResult<Node> {
        self.parse_expr()
    }

    pub fn parse_top_level(&mut self) -> LamResult<Node> {  self.parse_primary()  }

    pub(crate) fn got_tokens(&self) -> bool {
        self.pos < self.tokens.len()
    }

    fn parse_expr(&mut self) -> LamResult<Node> {
        let mut node = self.parse_primary()?;

        while self.do_primary() {
            let arg = self.parse_primary()?;
            node = Node::Application { func: Box::from(node), arg: Box::new(arg) };
        }

        /* x |> y */
        while matches!(self.peek(), Some(Token::Symbol(s)) if s == "|>") {
            self.consume();

            let mut f = self.parse_primary()?;

            /* get partial application args e.g. `x |> filter (> 5)` = `filter (> 5) x` */
            while self.do_primary() && !matches!(self.peek(), Some(Token::Symbol(s)) if s == "|>") {
                let a = self.parse_primary()?;
                f = Node::Application { func: Box::new(f), arg: Box::new(a) };
            }

            node = Node::Application { func: Box::new(f), arg: Box::new(node) };
        }

        Ok(node)
    }

    fn parse_primary(&mut self) -> LamResult<Node> {
        match self.consume() {
            Some(token) => match token {
                Token::LParen => {
                    match self.peek() {
                        Some(Token::Symbol(s)) if s == "let" => self.parse_let(),
                        Some(Token::Symbol(s)) if s == "fn" => self.parse_fn(),
                        Some(Token::Symbol(s)) if s == "if" => self.parse_if(),
                        Some(Token::Symbol(s)) if s == "match" => self.parse_match(),
                        Some(Token::Symbol(s)) if s == "use" => self.parse_use(),
                        Some(Token::Symbol(s)) if !s.chars().next().unwrap().is_ascii_alphabetic() => self.parse_partial(),
                        _ => {
                            let i = self.parse_expr()?;
                            match self.consume() {
                                Some(Token::RParen) => {},
                                t => return Err(LamError::new(format!("Expected ')' but got '{t:?}'"))),
                            }
                            Ok(i)
                        }
                    }
                },
                Token::Lambda => self.parse_lambda(),
                Token::LBracket => self.parse_list(),
                Token::Number(n) => Ok(Node::Literal(*n)),
                Token::Char(c) => Ok(Node::Literal(f64::from(*c as u32))),
                Token::Symbol(s) | Token::String(s) => Ok(Node::Atom(s.clone())),
                _ => Err(LamError::new(format!("Unexpected token '{token:?}'"))),
            }
            _ => Err(LamError::new("Unexpected end of input")),
        }
    }

    fn parse_partial(&mut self) -> LamResult<Node> {
        let op = match self.consume() {
            Some(Token::Symbol(s)) => s.clone(),
            _ => panic!("expected operator in partial")
        };

        let mut args = vec![];
        while !matches!(self.peek(), Some(Token::RParen)) {
            args.push(self.parse_primary()?);
        }

        self.consume();

        match args.len() {
            0 => Ok(Node::Atom(op)),
            1 => Ok(Node::Partial { op, arg: Box::new(args.pop().unwrap())}),
            2 => {
                let r = args.pop().unwrap();
                let l = args.pop().unwrap();
                Ok(Node::Application {
                    func: Box::new(Node::Partial { op, arg: Box::new(l) }),
                    arg: Box::new(r)
                })
            }
            _ => Err(LamError::new("Operation section takes one to two arguments"))
        }
    }

    fn parse_list(&mut self) -> LamResult<Node> {
        let mut list: Vec<Node> = vec![];

        if matches!(self.peek(), Some(Token::RBracket)) {
            self.consume();
            return Ok(Node::List(list));
        }

        let first_elem = self.parse_primary()?;

        /* [x..y] or [x..y;z] */
        if matches!(self.peek(), Some(Token::Range)) {
            self.consume();

            let to = self.parse_primary()?;
            let step = if matches!(self.peek(), Some(Token::Semi)) {
                self.consume();
                self.parse_primary()?
            } else {
                Node::Literal(1f64)
            };

            match self.consume() {
                Some(Token::RBracket) => {},
                _ => return Err(LamError::new("Expected ']' to close range")),
            }

            match (&first_elem, &to, &step) {
                (Node::Literal(f), Node::Literal(t), Node::Literal(s)) => {
                    let mut value = *f;
                    while value < *t {
                        list.push(Node::Literal(value));
                        value += s;
                    }
                    return Ok(Node::List(list));
                }
                _ => return Err(LamError::new("Range bounds must be numeric literals"))
            }
        }

        list.push(first_elem);

        loop {
            match self.consume() {
                Some(Token::RBracket) => return Ok(Node::List(list)),
                Some(Token::Comma) => list.push(self.parse_primary()?),
                _ => return Err(LamError::new("Expected ',' or ']' in list")),
            }
        }
    }

    fn parse_lambda(&mut self) -> LamResult<Node> {
        /* parse e.g. (\x -> (+ x 1)) */
        let mut params = vec![];
        while !matches!(self.peek(), Some(Token::Symbol(s)) if s == "->") {
            match self.consume() {
                Some(Token::Symbol(s)) => params.push(s.clone()),
                _ => return Err(LamError::new("Expected parameter name in lambda")),
            }
        }

        self.consume();

        let body = self.parse_primary()?;

        Ok(Node::LambdaDef { params, body: Box::new(body) })
    }

    fn parse_let(&mut self) -> LamResult<Node> {
        /* parse e.g. (let (x 5) (+ x 1)) */
        /* ( -> let -> (name value) -> body -> ) */
        self.consume();

        match self.consume() {
            Some(Token::LParen) => {},
            _ => return Err(LamError::new("Expected '(' after 'let'"))
        }

        let name = match self.consume() {
            Some(Token::Symbol(s)) => s.clone(),
            _ => return Err(LamError::new("Expected variable name in let"))
        };

        let value = self.parse_expr()?;

        match self.consume() {
            Some(Token::RParen) => {},
            _ => return Err(LamError::new("Expected ')' to close let binding"))
        }

        let body = self.parse_expr()?;

        match self.consume() {
            Some(Token::RParen) => {},
            _ => return Err(LamError::new("Expected ')' to close let")),
        }

        Ok(Node::Let { name, value: Box::new(value), body: Box::new(body) })
    }

    fn parse_fn(&mut self) -> LamResult<Node> {
        /* parse e.g. (fn (double x) ((* 2) x)) */
        /* ( -> fn -> (name params) -> body -> ) */
        self.consume();

        match self.consume() {
            Some(Token::LParen) => {},
            _ => return Err(LamError::new("Expected '(' after 'fn'")),
        }

        let name = match self.consume() {
            Some(Token::Symbol(s)) => s.clone(),
            _ => return Err(LamError::new("Expected function name")),
        };

        let mut params = vec![];
        while !matches!(self.peek(), Some(Token::RParen)) {
            match self.consume() {
                Some(Token::Symbol(s)) => params.push(s.clone()),
                _ => return Err(LamError::new("Expected parameter name")),
            }
        }

        self.consume();

        let body = self.parse_expr()?;

        match self.consume() {
            Some(Token::RParen) => {},
            _ => return Err(LamError::new("Expected ')' to close fn")),
        }

        Ok(Node::FnDef { name, params, body: Box::new(body) })
    }

    fn parse_if(&mut self) -> LamResult<Node> {
        self.consume();

        let cond = self.parse_primary()?;
        let then = self.parse_primary()?;
        let els = self.parse_primary()?;

        match self.consume() {
            Some(Token::RParen) => {},
            _ => return Err(LamError::new("Expected ')' to close if")),
        }

        Ok(Node::If { cond: Box::new(cond), then: Box::new(then), els: Box::new(els) })
    }

    fn parse_match(&mut self) -> LamResult<Node> {
        self.consume();

        let scrutinee = self.parse_primary()?;
        let mut arms = vec![];

        while matches!(self.peek(), Some(Token::LParen)) {
            self.consume();
            let pattern = self.parse_pattern()?;
            let guard = if matches!(self.peek(), Some(Token::Symbol(s)) if s == "if") {
                self.consume();
                Some(Box::new(self.parse_primary()?))
            } else {
                None
            };

            let body = self.parse_primary()?;
            match self.consume() {
                Some(Token::RParen) => {},
                _ => return Err(LamError::new("Expected ')' to close match arm")),
            }

            arms.push(MatchArm { pattern, guard, body: Box::new(body) });
        }

        match self.consume() {
            Some(Token::RParen) => {},
            _ => return Err(LamError::new("Expected ')' to close match")),
        }

        Ok(Node::Match { expr: Box::new(scrutinee), arms })
    }

    fn parse_use(&mut self) -> LamResult<Node> {
        self.consume();
        let path = match self.consume() {
            Some(Token::String(s)) => s.clone(),
            _ => return Err(LamError::new("Expected a string in 'use'")),
        };
        match self.consume() {
            Some(Token::RParen) => {},
            _ => return Err(LamError::new("Expected ')' to close use")),
        }

        Ok(Node::UseModule { path })
    }

    fn parse_pattern(&mut self) -> LamResult<Pattern> {
        match self.consume() {
            Some(Token::Number(n)) => Ok(Pattern::Literal(*n)),
            Some(Token::String(s)) => Ok(Pattern::StringLit(s.clone())),
            Some(Token::Symbol(s)) if s == "_" => Ok(Pattern::Wildcard),
            Some(Token::Symbol(s)) => Ok(Pattern::Var(s.clone())),
            Some(Token::LBracket) => self.parse_list_pattern(),
            _ => Err(LamError::new("Unexpected token in pattern.")),
        }
    }

    fn parse_list_pattern(&mut self) -> LamResult<Pattern> {
        // println!("list pattern tokens: {&self.tokens[self.pos..]:?}");
        let mut patterns = vec![];
        let mut rest = None;

        if matches!(self.peek(), Some(Token::RBracket)) {
            self.consume();
            return Ok(Pattern::List(patterns, rest));
        }

        loop {
            if matches!(self.peek(), Some(Token::Spread)) {
                self.consume();
                match self.consume() {
                    Some(Token::Symbol(s)) => rest = Some(s.clone()),
                    _ => return Err(LamError::new("Expected name after '...' (spread)")),
                }
            } else {
                patterns.push(self.parse_pattern()?);
            }

            match self.consume() {
                Some(Token::RBracket) => return Ok(Pattern::List(patterns, rest)),
                Some(Token::Comma) => {},
                _ => return Err(LamError::new("Expected ',' or ']' in list pattern.")),
            }
        }
    }

    fn peek(&self) -> Option<&Token> {
        self.tokens.get(self.pos)
    }

    fn consume(&mut self) -> Option<&Token> {
        let t = self.tokens.get(self.pos);
        self.pos += 1;
        t
    }

    fn do_primary(&self) -> bool {
        match self.peek() {
            Some(Token::RParen | Token::RBracket | Token::Comma) => false,
            Some(Token::Symbol(s)) if s == "|>" => false,
            Some(Token::LParen) => {
                !matches!(self.tokens.get(self.pos + 1),
                    Some(Token::Symbol(s)) if matches!(s.as_str(), "fn" | "let" | "if"))
            }
            Some(_) => true,
            None => false,
        }
    }
}

fn tokenize(input: &str) -> Vec<Token> {
    let mut tokens = vec![];
    let mut chars = input.char_indices().peekable();

    while let Some((pos, current)) = chars.next() {
        match current {
            c if c.is_ascii_whitespace() => {},
            '(' => tokens.push(Token::LParen),
            ')' => tokens.push(Token::RParen),
            '[' => tokens.push(Token::LBracket),
            ']' => tokens.push(Token::RBracket),
            ',' => tokens.push(Token::Comma),
            ';' => tokens.push(Token::Semi),
            '.' if chars.peek().is_some_and(|(_, c)| *c == '.') => {
                chars.next();
                if chars.peek().is_some_and(|(_, c)| *c == '.') {
                    chars.next();
                    tokens.push(Token::Spread);
                } else {
                    tokens.push(Token::Range);
                }
            }
            '\\' => tokens.push(Token::Lambda),
            n if n.is_ascii_digit() => {
                let start = pos;
                let mut is_flt = false;

                while chars.peek().is_some_and(|(p, c)| {
                    if c.is_ascii_digit() {
                        true
                    } else if *c == '.' && !is_flt {
                        /* had the issue of ".." being a float, so uh set is_flt ONLY if peeked next char is a number */
                        let nx = input.as_bytes().get(p + 1);
                        if nx.is_some_and(u8::is_ascii_digit) {
                            is_flt = true;
                            true
                        } else {
                            false
                        }
                    } else {
                        false
                    }
                }) {
                    chars.next();
                }

                let end = chars.peek().map_or(input.len(), |(i, _)| *i);
                let value = &input[start..end];

                tokens.push(Token::Number(value.parse::<f64>().unwrap_or_else(|_| panic!("failed to parse '{value}' as number"))));
            },
            q if q == '"' => {
                let start = pos + 1;

                while chars.peek().is_some_and(|(_, c)| *c != q) {
                    chars.next();
                }

                let end = chars.peek().map_or(input.len(), |(i, _)| *i);
                chars.next();

                tokens.push(Token::String(input[start..end].to_string()));
            },
            '\'' => {
                if let Some((_, chr)) = chars.next() {
                    if chars.next().is_some_and(|(_, nxt)| nxt == '\'') {
                        tokens.push(Token::Char(chr));
                    } else if chars.next().is_some_and(|(_, nxt)| nxt.is_ascii_alphanumeric()) {
                        panic!("char literals can only contain one character");
                    } else {
                        panic!("unterminated char literal");
                    }
                }
            },
            _ => {
                let start = pos;

                while chars.peek().is_some_and(|(_, c)| !is_delimiter(*c)) {
                    chars.next();
                }

                let end = chars.peek().map_or(input.len(), |(i, _)| *i);
                let value = &input[start..end];

                tokens.push(Token::Symbol(value.to_string()));
            },
        }
    }

    tokens
}

fn is_delimiter(c: char) -> bool {
    c.is_ascii_whitespace() || matches!(c, '(' | ')' | '[' | ']' | '\'' | '"' | ';' | '.' | ',')
}